### PR TITLE
init: Pebble sdk at 4.5

### PR DESCRIPTION
**Pull request by edanaher**
###### Motivation for this change

While Pebble has been bought by Fitbit and stopped selling new products, the watches are still around and available (and cheap!). So this adds the pebble sdk to develop apps for it.

After the initial PR (#23445) languished and was eventually closed, here's a fresh one.  (It's been a good learning experience for me in the Right Way to do several things :) )

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -16241,6 +16272,24 @@ in {
       };
     in if isPy3k then py3 else py2;
 
+  pyqrcode = buildPythonPackage rec {
+    name = "${pname}-${version}";
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: No `name`.

File: pkgs/top-level/python-packages.nix [hunk 2]
```diff
@@ -18240,6 +18289,20 @@ EOF
 
   magic-wormhole = callPackage ../development/python-modules/magic-wormhole { };
 
+  wsgiref = buildPythonPackage rec {
+    name = "wsgiref-0.1.2";
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: No `name`. Specify `pname` and `version` instead.

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -18240,6 +18289,20 @@ EOF
...
+    src = pkgs.fetchurl {
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: `fetchPypi`

File: pkgs/top-level/python-packages.nix [hunk 4]
```diff
@@ -18240,6 +18289,20 @@ EOF
...
+      url = "mirror://pypi/w/wsgiref/${name}.zip";
+      sha256 = "0y8fyjmpq7vwwm4x732w97qbkw78rjwal5409k04cw4m03411rn7";
+    };
+
+    meta = with pkgs.stdenv.lib; {
+      homepage = https://pypi.python.org/pypi/wsgiref;
+      license = "PSF or ZPL";
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: You can use a list of licenses.

File: pkgs/top-level/python-packages.nix [hunk 5]
```diff
@@ -16241,6 +16272,24 @@ in {
...
+
+    src = fetchPypi {
+      inherit pname version;
+      sha256 = "1m9ln8k9v7dfbh1i81225hx5mdsh8mpf9g7r4wpbfmiyfcs7dgzx";
+    };
+
+    meta = {
+      description = "A QR code generator written purely in Python with SVG, EPS, PNG and terminal output";
+      home = "https://pypi.python.org/pypi/PyQRCode/";
+      license = "bsd";
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: Which of the BSDs? Please use an attribute from https://github.com/NixOS/nixpkgs/blob/master/lib/licenses.nix

File: pkgs/development/mobile/pebblesdk/pebble-libs.nix [hunk 6]
```diff
@@ -0,0 +1,59 @@
+{ pkgs, python27Packages }:
+
+rec {
+  libpebble2 = python27Packages.buildPythonPackage rec {
+    name = "${pname}-${version}";
```
Comment (dotlambda) [inline pkgs/development/mobile/pebblesdk/pebble-libs.nix]: No `name`.

File: pkgs/development/mobile/pebblesdk/pebble-libs.nix [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+      sha256 = "16n69xxma7k8mhl8birdwa0fsqvf902g08s80mjb477s4dcxrvaz";
+    };
+
+    # Tries and fails to pull in pytest-mock
+    doCheck = false;
+
+    propagatedBuildInputs = with python27Packages; [ enum34 pyserial six websocket_client ];
+
+    meta = with pkgs.stdenv.lib; {
+      homepage = "";
```
Comment (dotlambda) [inline pkgs/development/mobile/pebblesdk/pebble-libs.nix]: If there is no homepage, please leave it out.

File: pkgs/development/mobile/pebblesdk/pebble-libs.nix [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+      license = licenses.mit;
+      description = "Library for communicating with pebbles over pebble protocol";
+    };
+  };
+
+  pypkjs = python27Packages.buildPythonPackage {
+    name = "pypkjs-1.0.6";
```
Comment (dotlambda) [inline pkgs/development/mobile/pebblesdk/pebble-libs.nix]: No `name`. Specify `pname` and `version` instead.

File: pkgs/development/mobile/pebblesdk/pebble-libs.nix [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+      pypng
+      dateutil
+      requests
+      sh
+      six
+      websocket_client
+      wsgiref
+    ];
+
+    patchPhase = ''
```
Comment (dotlambda) [inline pkgs/development/mobile/pebblesdk/pebble-libs.nix]: `prePatch`

File: pkgs/development/mobile/pebblesdk/pebble-libs.nix [hunk 10]
```diff
@@ -0,0 +1,59 @@
+{ pkgs, python27Packages }:
+
+rec {
+  libpebble2 = python27Packages.buildPythonPackage rec {
+    name = "${pname}-${version}";
+    pname = "libpebble2";
+    version = "0.0.26";
+    src = python27Packages.fetchPypi {
```
Comment (dotlambda) [inline pkgs/development/mobile/pebblesdk/pebble-libs.nix]: If this is even on PyPI, why didn't you put it in `python-packages.nix`?

File: pkgs/development/mobile/pebblesdk/pebble-libs.nix [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+      inherit pname version;
+      sha256 = "16n69xxma7k8mhl8birdwa0fsqvf902g08s80mjb477s4dcxrvaz";
+    };
+
+    # Tries and fails to pull in pytest-mock
```
Comment (dotlambda) [inline pkgs/development/mobile/pebblesdk/pebble-libs.nix]: `checkInputs = [ pytest-mock ]`

File: pkgs/development/mobile/pebblesdk/default.nix [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+     --set PEBBLE_TOOLCHAIN_PATH $out/arm-cs-tools/bin \
+     --set PHONESIM_PATH ${pebble-libs.pypkjs}/bin/pypkjs \
+     --set PATH $PATH:${nodejs}/bin
+  '';
+
+  meta = with stdenv.lib; {
+    homepage = https://developer.pebble.com/sdk;
+    description = "SDK to build applications for the Pebble smartwatch";
+
+    license = https://developer.pebble.com/legal/sdk-license/;
```
Comment (dotlambda) [inline pkgs/development/mobile/pebblesdk/default.nix]: This should probably be `licenses.unfree`, shouldn't it?
You can put the URL in a comment.

File: pkgs/top-level/python-packages.nix [hunk 13]
```diff
@@ -16241,6 +16272,24 @@ in {
...
+    version = "1.2.1";
+
+    src = fetchPypi {
+      inherit pname version;
+      sha256 = "1m9ln8k9v7dfbh1i81225hx5mdsh8mpf9g7r4wpbfmiyfcs7dgzx";
+    };
+
+    meta = {
+      description = "A QR code generator written purely in Python with SVG, EPS, PNG and terminal output";
+      home = "https://pypi.python.org/pypi/PyQRCode/";
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: `homepage`, no quotes. And the homepage should be https://github.com/mnooner256/pyqrcode.

**Review by dotlambda — CHANGES_REQUESTED** [review]

**Review by dotlambda — CHANGES_REQUESTED** [review]
Please also make all changes requested in https://github.com/NixOS/nixpkgs/pull/23445.

**Comment by edanaher** [comment]
*sigh*; I forgot to push my final two changes wrapping up comments from #23445.  Those should be good now; I'll look at the rest of your comments in a bit.

**Comment by edanaher** [comment]
@dotlambda Thanks for all the feedback; I think I've addressed it all, along with a couple other minor cleanups.  Two areas I'd appreciate feedback:

- It looks like libpebble and peewee don't put their tests in the pypi distribution; I'm not entirely sure whether it's better to be simple and disable the tests, or pull the source from github and enable them

- progressbar2 version 3 made some breaking changes, so I'm building progressbar2 version 2.7.3 (the latest 2.x version I could find).  I'm not sure if that should be in pebble-libs or if I should add it to python-packages in case other packages also need it.

**Comment by dotlambda** [comment]
- Better fetch from GitHub and, if you like, raise an upstream issue about including tests on PyPI.
- I think you should add `libpebble2` to the python package set (using `callPackage` from `python-packages.nix`). You can then override the progressbar2 version like so: https://github.com/peterhoeg/nixpkgs/blob/master/pkgs/servers/home-assistant/default.nix#L8 and create a `pypkjs.nix` file which you call just like `frontend.nix` in the example.

**Comment by dotlambda** [comment]
@edanaher Any progress?

**Comment by edanaher** [comment]
In the process of pulling out and testing pypkjs, I discovered that the emulator isn't working again; I was looking at that, and got distracted.  I'll take another shot at fixing it up.

**Pull request closed without merging**